src/plugin/protocol-http/src/java/org/apache/nutch/protocol/http/HttpResponse.java

```diff
@@ -193,34 +197,54 @@ public HttpResponse(HttpBase http, URL url, CrawlDatum datum)
       reqStr.append("\r\n");
 
       if (http.isIfModifiedSinceEnabled() && datum.getModifiedTime() > 0) {
-        reqStr.append("If-Modified-Since: "
-            + HttpDateFormat.toString(datum.getModifiedTime()));
+        reqStr.append("If-Modified-Since: " + HttpDateFormat
+            .toString(datum.getModifiedTime()));
         reqStr.append("\r\n");
       }
       reqStr.append("\r\n");
 
+      // store the request in the metadata?
+      if (conf.getBoolean("store.http.request", false) == true) {
+        headers.add("_request_", reqStr.toString());
+      }
+
       byte[] reqBytes = reqStr.toString().getBytes();
 
       req.write(reqBytes);
       req.flush();
 
       PushbackInputStream in = // process response
-      new PushbackInputStream(new BufferedInputStream(socket.getInputStream(),
-          Http.BUFFER_SIZE), Http.BUFFER_SIZE);
+          new PushbackInputStream(
+              new BufferedInputStream(socket.getInputStream(),
+                  Http.BUFFER_SIZE), Http.BUFFER_SIZE);
 
       StringBuffer line = new StringBuffer();
 
+      // store the http headers verbatim
+      if (conf.getBoolean("store.http.headers", false) == true) {
+        httpHeaders = new StringBuffer();
+      }
+
+      headers.add("nutch.fetch.time", Long.toString(datum.getFetchTime()));
```

Comment:
Why take `datum.getFetchTime()` and what does its value correspond to? Isn't that the time is was due for fetching?
I'd have used `System.currentTimeMillis()` as this is exactly the moment we fetch the content and would definitely be accurate